libdnf/base/goal.cpp

```diff
@@ -596,10 +604,136 @@ void Goal::Impl::add_rpms_to_goal(base::Transaction & transaction) {
                     settings.resolve_best(cfg_main),
                     settings.resolve_clean_requirements_on_remove());
                 break;
+            case GoalAction::REINSTALL: {
+                solv::IdQueue ids_nevra_installed;
+                for (auto id : ids) {
+                    rpm::PackageQuery query(installed);
+                    query.filter_nevra({pool.get_nevra(id)});
+                    if (query.empty()) {
+                        // Report when package with the same NEVRA is not installed
+                        transaction.p_impl->add_resolve_log(
+                            GoalAction::REINSTALL,
+                            GoalProblem::NOT_INSTALLED,
+                            settings,
+                            {pool.get_nevra(id)},
+                            {},
+                            false);
+                    } else {
+                        // Only installed packages can be reinstalled
+                        ids_nevra_installed.push_back(id);
+                    }
+                }
+                rpm_goal.add_install(
+                    ids_nevra_installed,
+                    settings.resolve_strict(cfg_main),
+                    settings.resolve_best(cfg_main),
+                    settings.resolve_clean_requirements_on_remove());
+            } break;
             case GoalAction::UPGRADE: {
+                solv::IdQueue ids_upgrades;
+                for (auto id : ids) {
+                    rpm::PackageQuery query(installed);
+                    query.filter_name({pool.get_name(id)});
```

Comment:
Before it starts to evaluate anything we have to test wheather the package is obsoleter. See dnf base.py line 2140
```
                obsoleters = q.filter(obsoletes=installed) \
                    if self.conf.obsoletes else self.sack.query().filterm(empty=True)
                if not obsoleters:
```

Reply:
Ok. I added support for obsoleters.